gui/gui_state_container.h

```diff
@@ -0,0 +1,42 @@
+#ifndef GUI_STATE_CONTAINER_H
+#define GUI_STATE_CONTAINER_H
+
+#include "controller_history.h"
+#include <chrono>
+#include <mutex>
+
+// A thread-safe container for readable and writable state
+// of the GUI.
+class GuiStateContainer {
```

Comment:
I don't get the purpose of this class -- right now it seems to be just a mutex wrapped around the ControllerHistory.  There's also a GuiStatus in here, but it's lacking mutators (add a TODO?).

I guess my question is, does there need to be one lock guarding the GuiStatus and history objects?  If not, could we make the history object thread-safe and then get rid of this entirely?  Or is going somewhere that I'm not envisioning.

Reply:
I prefer to have as few components as possible claim to be thread-safe (=> less stuff can break and less stuff has to be tested), so I didn't want to make ControllerHistory thread-safe - instead I made this container with a dumb mutex. I figured it will be easy for people to modify this with mutators (yes, will add a TODO) without breaking thread-safety. It also appealed to me to have a central class containing the readable and writable state of the GUI, because it resembles an MVC "model".

Reply:
OK, I don't quite see how this is all going to come together, but it sounds like you have a plan.  Maybe could we write down a sentence about the design philosophy?

Reply:
Added some comments.